reconcile/terraform_repo.py

```diff
@@ -0,0 +1,304 @@
+    repo_list: list[TFRepo] = []
+    keys = state.ls()
+    for key in keys:
+        value = state.get(key.lstrip("/"), None)
+        repo = TFRepo.parse_raw(value)
+
+        if repo is not None:
+            repo_list.append(repo)
+
+    return repo_list
+
+
+def map_repos(repos: list[TFRepo]) -> MutableMapping[str, TFRepo]:
+    """Generate keys for each repo to more easily compare"""
+    repo_map: MutableMapping[str, TFRepo] = dict()
+    for repo in repos:
+        # repo names are unique per app interface instance as enforced in the schema
+        repo_map[repo.name] = repo
+
+    return repo_map
+
+
+def check_ref(repo_url: str, ref: str, path: str) -> None:
+    """
+    Checks whether a Git ref is valid
+    and whether config.tf exists in the project path"""
+    instance = queries.get_gitlab_instance()
+    with GitLabApi(
+        instance, settings=queries.get_secret_reader_settings(), project_url=repo_url
+    ) as gl:
+        try:
+            gl.get_commit_sha(ref=ref, repo_url=repo_url)
+            retrieved_path = gl.get_file("{}/config.tf".format(path), ref)
+
+            if not retrieved_path:
+                raise ParameterError(
+                    'No config.tf found in path: "{}" on repo: "{}". Ensure that a config.tf file is present in this path.'.format(
+                        path, repo_url
+                    )
+                )
+        except (KeyError, AttributeError):
+            raise ParameterError(
+                'Invalid ref: "{}" on repo: "{}"'.format(ref, repo_url)
+            )
+
+
+def diff_missing(
+    a: Mapping[str, TFRepo],
+    b: Mapping[str, TFRepo],
+    requireDelete: bool = False,
+) -> MutableMapping[str, TFRepo]:
+    """
+    Returns a mapping of repos that are present in mapping a but missing in mapping b
+    When requireDelete = True, each repo from a is checked to ensure that the delete flag is set"""
+    missing: MutableMapping[str, TFRepo] = dict()
+    for a_key, a_repo in a.items():
+        b_repo = b.get(a_key)
+        if b_repo is None:
+            if requireDelete is True and a_repo.delete is not True:
+                raise ParameterError(
+                    'To delete the terraform repo "{}", you must set delete: true in the repo definition'.format(
+                        a_repo.name
+                    )
+                )
+            missing[a_key] = a_repo
+
+    return missing
+
+
+def diff_changed(
+    a: Mapping[str, TFRepo], b: Mapping[str, TFRepo]
+) -> MutableMapping[str, TFRepo]:
+    """
+    Returns a mapping of repos that have changed between a and b in the form of returning their b values
+    In order for Terraform to work as expected, the tenant can only update the ref between MRs
+    Updating account, repo, or project_path can lead to unexpected behavior so we error
+    on that
+    """
+    changed: MutableMapping[str, TFRepo] = dict()
+    for a_key, a_repo in a.items():
+        b_repo = b.get(a_key, a_repo)
```

Comment:
if `b` has no `a_key` then this will return `a_repo` as value, is this expected? It will pass diff implicitly, maybe worth do a simple get query and None check to make it explicit.